atlasdb-client/src/main/java/com/palantir/atlasdb/cache/DefaultOffHeapCache.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cache;
+
+import static com.palantir.atlasdb.cache.OffHeapCaches.OFF_CACHE_NAMESPACE;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.Futures;
+import com.palantir.atlasdb.autobatch.Autobatchers;
+import com.palantir.atlasdb.autobatch.CoalescingRequestFunction;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.atlasdb.persistent.api.LogicalPersistentStore;
+import com.palantir.atlasdb.persistent.api.PhysicalPersistentStore.StoreNamespace;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+final class DefaultOffHeapCache<K, V> implements OffHeapCache<K, V> {
+    private static final Logger log = LoggerFactory.getLogger(DefaultOffHeapCache.class);
+    private static final String BATCHER_PURPOSE = "off-heap-cache";
+    private static final MetricName CACHE_HIT = constructCacheMetricName("cacheHit");
+    private static final MetricName CACHE_MISS = constructCacheMetricName("cacheMiss");
+    private static final MetricName CACHE_NUKE = constructCacheMetricName("cacheNuke");
+    private static final MetricName CACHE_SIZE = constructCacheMetricName("cacheSize");
+
+    private final LogicalPersistentStore<K, V> persistentStore;
```

Comment:
as an aside, I don't understand why you need to distinguish between `Logical` and `Physical`, if anything, the latter is an instance of the former where `K` and `V` are both `ByteString`...

Reply:
Acked and removed.